src/tap/src/tapcell.cpp

```diff
@@ -57,17 +63,1091 @@ Tapcell::~Tapcell()
+
+  if (x < 0 || y < 0) {
+    return;
+  }
+  std::string name = prefix + std::to_string(Tapcell::phy_idx);
+  odb::dbInst* inst = odb::dbInst::create(block, master, name.c_str());
+  if (inst == nullptr) {
+    return;
+  }
+  inst->setOrient(orientation);
+  inst->setLocation(x, y);
+  inst->setPlacementStatus(odb::dbPlacementStatus::LOCKED);
+  inst->setSourceType(odb::dbSourceType::DIST);
+
+  odb::dbBox* inst_bb = inst->getBBox();
+  std::vector<int> filled_sites_vector;
+  filled_sites_vector.push_back(inst_bb->yMin());
+  filled_sites_vector.push_back(inst_bb->xMin());
+  filled_sites_vector.push_back(inst_bb->xMax());
+  Tapcell::filled_sites.push_back(filled_sites_vector);
+
+  Tapcell::phy_idx++;
+}
+
+void Tapcell::buildRow(odb::dbBlock* block, std::string name, odb::dbSite* site, int start_x, int end_x, int y, odb::dbOrientType& orient, odb::dbRowDir& direction, int& min_row_width) {
+  int site_width = site->getWidth();
+
+  int new_row_num_sites = (end_x - start_x)/site_width;
+  int new_row_width = new_row_num_sites*site_width;
+
+  if (new_row_num_sites > 0 && new_row_width >= min_row_width) {
+    odb::dbRow::create(block, name.c_str(), site, start_x, y, orient, direction, new_row_num_sites, site_width);
+  }
+}
+
+std::vector<std::vector<odb::dbRow*>> Tapcell::organizeRows() {
+  std::map<int, std::vector<odb::dbRow*>> rows_dict;
+  for (odb::dbRow* row : db_->getChip()->getBlock()->getRows()) {
+    odb::Rect rowBB;
+    row->getBBox(rowBB);
+    if(rows_dict.find(rowBB.yMin())==rows_dict.end()) {
+      std::vector<odb::dbRow*> rows_vector;
+      rows_vector.push_back(row);
+      rows_dict.insert(std::pair<int, std::vector<odb::dbRow*>>(rowBB.yMin(), rows_vector));
+    } else {
+      rows_dict[rowBB.yMin()].push_back(row);
+    }
+  }
+
+  // organize rows bottom to top
+  std::vector<std::vector<odb::dbRow*>> rows;
+  for(std::map<int, std::vector<odb::dbRow*>>::iterator iter = rows_dict.begin(); iter != rows_dict.end(); ++iter)
+  {
+    int k =  iter->first;
+    std::map<int, odb::dbRow*> in_row_dict;
+    for(odb::dbRow* in_row : rows_dict[k]) {
+      odb::Rect row_BB;
+      in_row->getBBox(row_BB);
+      in_row_dict.insert(std::pair<int, odb::dbRow*>(row_BB.xMin(), in_row));
+    }
+    // organize sub rows left to right
+    std::vector<odb::dbRow*> in_row;
+    for(std::map<int, odb::dbRow*>::iterator iter = in_row_dict.begin(); iter != in_row_dict.end(); ++iter)
+    { 
+      int in_k =  iter->first;
+      in_row.push_back(in_row_dict[in_k]);
+    }
+    rows.push_back(in_row);
+    in_row.clear();
+  }
+  rows_dict.clear();
+
+  return rows;
+}
+
+int Tapcell::removeCells(const char* prefix) {
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  std::string prefix_string(prefix);
+  int removed = 0;
+
+  if (prefix_string.length() == 0) {
+    // no prefix is given, this will result in all cells being removed
```

Comment:
It doesn't look like any cells are removed

Reply:
I am not sure honestly how the cells are removed in that situation. @eder-matheus is the comment wrong or should I add something other than what's in the tcl code to remove all cells?

Reply:
I think this was added in the tcl code to prevent the following loop from deleting all the cells. Without it calling this function without a prefix defined would result in all instances getting deleted.